Review treatment of issued at, issuer, audience #16

dknoma opened the pull request:
So far I implemented newer versions of `resolve`, `lookup`, `unresolve`, and `OAuthRealm`. They may require other parameters to include the issuer and audience, therefor making major changes to the code. Rather than just a `Map<String, OAuthRealm>` it’s now a `Map<String, List<OAuthRealm>>` where a helper function will go through the list to filter out the realms with the correct issuer and audience, going off of our talk of ways to determine realm bits depending on the existence of issuer and audience. 

All tests are passing and builds correctly, tho I did have to remove all the `.claim("iss", "test issuer")` from the spec scripts and created a script that explicity uses `iss` and `aud` to make it explicit that these are testing the new claims rather than having them in ALL the scripts.

dknoma committed:
added a new test to check for issuer and audience claims. OAuthResolveExFW gets generated correctly, and is able to wrap the original ResolveFW to get the issuer and audience claims from it.

dknoma committed:
testing new methods to allow for iss and aud claims. most spec scripts use "iss": "test issuer" in their jwt so either their routes must also include that claim, or get rid of those claims all together if we want to only test those claims in a few specific tests

dknoma committed:
removed `iss` claims from scripts that do not test for issuer and audience claims. all tests passing

dknoma committed:
fixed `unresolve()` to return the correct value

dknoma committed:
removed unnecessary code from filtered search of the `List<OAuthRealm>`

dknoma committed:
removed wildcard import

dknoma committed:
removed wildcard flyweight imports in OAuthNukleus

dknoma committed:
fixed order of flyweight imports

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -74,39 +79,73 @@ private OAuthRealms(
+            final List<OAuthRealm> realms = realmsIdsByName.computeIfAbsent(realmName, r -> new LinkedList<>());
+            final OAuthRealm realm = getNewRealmIfAbsent(realmName, realms, issuerName, audienceName);
             authorization = realm.resolve(scopeNames);
+            realmsIdsByName.get(realmName).add(realm);
```
Same as `realms.add(realm)` right?
However, if seems like this code might add the same `realm` multiple times.

jfallows reviewed (changes requested):
Consider adding a class called `OAuthRealmInfo` to model the required values for `issuer` and `audience`.
```
OAuthRealms
    name (kid) -> OAuthRealm

OAuthRealm
    name (kid)
    list<OAuthRealmInfo>

OAuthRealmInfo
    realm bit
    claims (required claims; e.g. issuer, audience; possibly empty)
    map<role (scope), scope bit>
```

Then resolve would:
 - find or create the OAuthRealm within OAuthRealms based on the realm name
 - find or create the OAuthRealmInfo within OAuthRealm based on the required claims
 - find or create the scope bits within OAuthRealmInfo based on the scope claim
 - return the authorization bits = realm bit | scope bits

This gives a distinct realm bit for different variations of enforced issuer or audience.

Then lookup would:
 - find the OAuthRealm by realm name (kid)
 - find the OAuthRealmInfo by required claims; issuer and audience
 - find the scope bits based on scope claim

This approach treats the absence of `issuer` or `audience` as if they are explicitly specified as `null`, meaning the if a route is setup as authorized without `issuer` or `audience`, then it is as if the route is setup with a `null` `issuer` or `null` audience, so if a JWT token is presented with an explicit value for `issuer` or `audience`, then it will _not_ match a route setup without a specific `issuer` or `audience`.

This is a reasonable place to start, but we may want to think about allocating a some of the realm bits as flags indicating whether or not an `issuer` or `audience` was specified. That would allow the less specific route (without `issuer` or without `audience`) to be followed when a more specific token including (`issuer` or `audience`) is presented with the HTTP request.

dknoma committed:
added a new OAuthRealmInfo class to hold info on a realms id, claims, and scopebits. builds successfully. still needs more testing and design to figure out how we should go about flagging the required claims. ie should they set aside a bit as a flag for these claims?

dknoma committed:
updated OAuthNukleus. removed comments and TODOs

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -193,9 +220,8 @@ public JsonWebKey supplyKey(
         private static final int MAX_SCOPES = 48;
 
-        private final Map<String, Long> scopeBitsByName = new CopyOnWriteHashMap<>();
+        private final List<OAuthRealmInfo> realmInfoVariations = new LinkedList<>();
```
`CopyOnWriteHashMap` was a concurrent data structure, supporting multiple threads accessing in parallel.
Is that scenario still necessary?

jfallows replied:
Please rename `realmInfoVariations` to `realmInfos`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -205,35 +231,81 @@ private OAuthRealm(
+            String audienceName)
+        {
+            assert nextRealmBit < MAX_REALMS;
+            OAuthRealmInfo info = getInfoByFilter(issuerName, audienceName);
```
```
return realmInfos.stream()
                 .filter(r -> r.containsClaims(issuerName, audienceName)
                 .findFirst()
                 .or(() -> newRealmInfo(issuerName, audienceName))
```
```
private OAuthRealmInfo newRealmInfo(
    String issuerName,
    String audienceName)
{
    OAuthRealmInfo realmInfo = new OAuthRealmInfo(1L << nextRealmBit++ << MAX_SCOPES, issuerName, audienceName);
    realmInfos.add(realmInfo);
    return realmInfo;
}
```

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -120,7 +132,7 @@ public boolean unresolve(
     {
         final long realmId = authorization & REALM_MASK;
-        return Long.bitCount(realmId) <= 1 && realmsIdsByName.entrySet().removeIf(e -> e.getValue().realmId == realmId);
+        return Long.bitCount(realmId) <= 1 && tryRemoveRealmInfoById(realmId);
```
```
        return Long.bitCount(realmId) <= 1 &&
                realmsIdsByName.values().removeIf(r -> r.unresolve(realmId) && r.isEmpty());
```
In `OAuthRealm`,
```
    boolean unresolve(
        long realmId)
    {
        return realmInfos.removeIf(i -> i.realmId == realmId);
    }

    private boolean isEmpty()
    {
        return realmInfos.isEmpty();
    }
```

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -205,35 +231,81 @@ private OAuthRealm(
-            String[] scopeNames)
+            String[] scopeNames,
+            String issuerName,
+            String audienceName)
```
Recommend refactoring the order of these parameters to `issuerName`, `audienceName`, `scopeNames`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -205,35 +231,81 @@ private OAuthRealm(
                 for (int i = 0; i < scopeNames.length; i++)
                 {
-                    authorization |= scopeBitsByName.computeIfAbsent(scopeNames[i], this::assignScopeBit);
+                    authorization |= realmInfo.computeScopeBitsIfAbsent(scopeNames[i], this::assignScopeBit);
```
Now that the scopes are being managed inside the `OAuthRealmInfo`, the supplier method `this::assignScopeBit` should also be moved inside `OAuthRealmInfo`, leaving this as:
```
authorization |= realmInfo.supplyScopeBit(scopeNames[i]);
```

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -205,35 +231,81 @@ private OAuthRealm(
-            String[] scopeNames)
+            String[] scopeNames,
+            String issuerName,
+            String audienceName)
```
Recommend refactoring the order of these parameters to `issuerName`, `audienceName`, `scopeNames`.

dknoma committed:
Updated methods w/ change requests. fixed imports and scope of certain methods as well.

dknoma committed:
fixed `unresolve` to correctly detect when an `OAuthRealmInfo` is removed from an `OAuthRealm`. in the corner case that a single kid has multiple unique sets of `issuer` and `audience`, a realm could before return false in `unresolve` even though an `info` object was removed from its list because it was checking if its list was empty to remove the realm. this means that the realm would not be removed therefore falsely returning false. Now filters the correct realm out via `OAuthRealm.unresolve()`, and gets a non-null `realm` if an `info` was removed successfully. It also correctly removes a `realm` from the map if its list is empty while still correctly returning the appropriate boolean value.

dknoma committed:
Merge remote-tracking branch 'upstream/develop' into develop

# Conflicts:
#	src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthElektron.java
#	src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java
#	src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java
#	src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactoryBuilder.java

dknoma committed:
merged upstream/develop into this fork. up to date now

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthNukleus.java:
```diff
@@ -40,6 +41,7 @@
     private static final String[] EMPTY_STRING_ARRAY = new String[0];
 
     private final ResolveFW resolveRO = new ResolveFW();
+    private final OAuthResolveExFW resolveROEx = new OAuthResolveExFW();
```
Rename `resolveROEx` to `resolveExRO`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthNukleus.java:
```diff
@@ -101,10 +103,20 @@ private void onResolve(
+        if(resolve.extension().sizeof() > 0)
+        {
+            final OAuthResolveExFW resolveExtension =
+                    resolveROEx.wrap(buffer, resolve.extension().offset(), resolve.extension().limit());
```
Consider using `tryWrap` instead of `wrap` to avoid exception when input is malformed, rename `resolveExtension` to `resolveEx` and handle case where `resolveEx` is `null`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -121,7 +130,11 @@ public boolean unresolve(
-        return Long.bitCount(realmId) <= 1 && realmsByName.entrySet().removeIf(e -> e.getValue().realmId == realmId);
+        final OAuthRealm realm = realmsByName.values().stream()
+                                                         .filter(rs -> rs.unresolve(realmId))
+                                                         .findFirst().orElse(null);
```
Move `orElse` to next line please.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -121,7 +130,11 @@ public boolean unresolve(
+        final OAuthRealm realm = realmsByName.values().stream()
+                                                         .filter(rs -> rs.unresolve(realmId))
+                                                         .findFirst().orElse(null);
+        realmsByName.values().removeIf(OAuthRealm::isEmpty);
```
Hoist `realmsByName.values()` into `realms` variable as it is used twice.
Excellent use of the `OAuthRealm::isEmpty` method reference!

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -194,58 +207,179 @@ public JsonWebKey lookupKey(
         private final String realmName;
 
-        private long nextScopeBit;
+        private int nextScopeBit = 0;
```
Good decision to use narrower `int` type instead of `long`.
Remove explicit assignment to `0`, this is automatic by Java spec.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -194,58 +207,179 @@ public JsonWebKey lookupKey(
+            if (nextScopeBit + scopeNames.length < MAX_SCOPES)
             {
-                authorization = realmId;
+                final OAuthRealmInfo realmInfo = computeInfoIfAbsent(issuerName, audienceName);
```
Inline `computeInfoIfAbsent` here.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -194,58 +207,179 @@ public JsonWebKey lookupKey(
         {
-            long authorization = realmId;
-            for (int i = 0; i < scopeNames.length; i++)
+            final OAuthRealmInfo info = getInfoByFilter(issuerName, audienceName);
```
Inline `getInfoByFilter` method here.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -194,58 +207,179 @@ public JsonWebKey lookupKey(
+                authorization = info.realmId;
+                for (int i = 0; i < scopeNames.length; i++)
+                {
+                    authorization |= info.getOrDefault(scopeNames[i], 0L);
```
Default is always `0L` so move that logic inside `getOrDefault` method and rename `getOrDefault` to `long scopeBit(String)`.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java:
```diff
@@ -481,7 +481,8 @@ private JsonWebSignature verifiedSignature(
                     final NumericDate notBefore = claims.getNotBefore();
-
+                    // TODO: put issued at claim check here?
+                    // final NumericDate issuedAt = claims.getIssuedAt();
```
Please remove these comments from this PR.

jfallows commented on src/test/java/org/reaktivity/nukleus/oauth/internal/control/ControlIT.java:
```diff
@@ -59,14 +59,14 @@ public void shouldFailToResolveWithTooManyRoles() throws Exception
+//    public void shouldFailToResolveWithTooManyRealms() throws Exception
+//    {
+//        k3po.finish();
+//    }
```
Uncomment?

dknoma commented:
I had to create a separate class, `ControlTooManyRealmsIT.java`, with a different ruleset because the configuration of the reaktors are different. But I successfully created another file with the test realms specifically for the too many realms tests and all tests are passing.

dknoma committed:
merged methods together that were not necessary to be kept separate.
fixed tests and created a new file to allow for too many realms tests
to function correctly. all tests passing and build is successful. Just
more general fine tuning and efficiency.

dknoma committed:
working on `OAuthController.resolve()`. successfully takes issuer and
audience names as parameters but fails to write out the correct message.

dknoma committed:
all tests work now. created another `resolve` method for `OAuthController`
which allows for the addition of issuer and audience claims for testing.

dknoma commented:
Removed the class previously mentioned and used the `@Configure` annotation to change which .jwk file is to be used for the test

dknoma committed:
created `newUnsafeBuffer` helper method as it will be used in multiple
methods and the constructor.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthController.java:
```diff
@@ -105,21 +101,18 @@ public String name()
         long correlationId = controllerSpi.nextCorrelationId();
-
-        OAuthResolveExFW resolveEx = resolveExRW.wrap(commandBuffer, 0, commandBuffer.capacity())
+        final UnsafeBuffer resolveExBuffer = newUnsafeBuffer();
```
Rename `resolveExBuffer` to `extensionBuffer` and avoid allocating on each call to `resolve` by hoisting to a `private final` member field alongside `commandBuffer`.

Inline the assignments to `commandBuffer` and `extensionBuffer` in constructor, no need for separate method here.

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthController.java:
```diff
@@ -105,21 +101,18 @@ public String name()
         String issuerName,
         String audienceName,
         String[] roles)
     {
```
Please match Parameter order to binary `RESOLVE` layout order; realmName, roles, issuerName, audienceName.

dknoma committed:
hoisted `extensionBuffer` up to the constructor and fixed the order of the parameters of `resolve`. fixed the spec scripts to support these extensions correctly

dknoma committed:
null extensions are treated as unspecified, and empty extensions are treated as any other non-null specified extension

dknoma committed:
update pom

jfallows commented on src/test/java/org/reaktivity/nukleus/oauth/internal/control/ControllerIT.java:
```diff
@@ -93,92 +95,95 @@ public void shouldFailToResolveWithTooManyRoles() throws Exception
+        thrown.expect(either(is(instanceOf(IllegalStateException.class)))
+                .or(is(instanceOf(ExecutionException.class))));
+        thrown.expectCause(either(nullValue(Exception.class)).or(is(instanceOf(IllegalStateException.class))));
+
```
Is it sufficient to remove these `thrown` lines and change annotation to `@Test( exception = IllegalStateException.class)` instead?
If so, that would be preferred as it is simpler.

dknoma replied:
I just tried it and it should work as `expected = ExecutionException.class`. I don't think this specific test will run into an `IllegalStateException` so I can go ahead and change that.

jfallows commented on src/test/java/org/reaktivity/nukleus/oauth/internal/control/ControllerIT.java:
```diff
@@ -236,6 +241,39 @@ public void shouldResolveWithRoles() throws Exception
+
+        long authorization = reaktor.controller(OAuthController.class)
+                .resolve("RS256", new String[]{"scope1", "scope2", "scope3"}, "test issuer", "test audience"
+                )
```
Close paren on wrong line.

dknoma committed:
changed unnecessary ExpectedException to an annotation. fixed wrong
closed parentheses.

dknoma committed:
fixed `shouldFailToResolveWithTooManyRealms` to expect the correct exception and cause

jfallows commented on src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:
```diff
@@ -194,58 +211,167 @@ public JsonWebKey lookupKey(
+            OAuthRealmInfo realmInfo =
+                    new OAuthRealmInfo(1L << nextRealmBit++ << MAX_SCOPES, issuerName, audienceName);
+//                            issuerName != null && !issuerName.isEmpty() ? issuerName : null,
+//                            audienceName != null && !audienceName.isEmpty() ? audienceName : null);
```
Remove commented code

dknoma committed:
removed commented code

jfallows committed:
Update nukleus-oauth.spec dependency version
